libs/MobileSync/src/com/salesforce/androidsdk/mobilesync/manager/AdvancedSyncUpTask.kt

```diff
@@ -0,0 +1,110 @@
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.salesforce.androidsdk.mobilesync.manager
+
+import com.salesforce.androidsdk.mobilesync.manager.SyncManager.SyncUpdateCallback
+import com.salesforce.androidsdk.mobilesync.target.AdvancedSyncUpTarget
+import com.salesforce.androidsdk.mobilesync.target.SyncUpTarget
+import com.salesforce.androidsdk.mobilesync.util.SyncOptions
+import com.salesforce.androidsdk.mobilesync.util.SyncState
+import com.salesforce.androidsdk.mobilesync.util.SyncState.MergeMode
+import com.salesforce.androidsdk.smartstore.store.SmartStore
+import org.json.JSONException
+import org.json.JSONObject
+import java.io.IOException
+
+/**
+ * Runnable class responsible for running a sync up that uses and AdvancedSyncUpTarget
+ */
+class AdvancedSyncUpTask(syncManager: SyncManager, sync: SyncState, callback: SyncUpdateCallback?) :
+    SyncUpTask(syncManager, sync, callback) {
+    @Throws(JSONException::class, IOException::class)
+    override fun syncUp(
+        sync: SyncState,
+        callback: SyncUpdateCallback?,
+        dirtyRecordIds: List<String>
+    ) {
+        val soupName = sync.soupName
+        val target = sync.target as SyncUpTarget
+        val options = sync.options
+        val totalSize = dirtyRecordIds.size
+        val maxBatchSize = (target as AdvancedSyncUpTarget).maxBatchSize
+        val batch: MutableList<JSONObject> = ArrayList()
+        updateSync(sync, SyncState.Status.RUNNING, 0, callback)
+        val dirtyRecords = target.getFromLocalStore(syncManager, soupName, dirtyRecordIds)
+
+        // Figuring out what records need to be synced up based on merge mode and last mod date on server
+        val recordIdToShouldSyncUp = shouldSyncUpRecords(syncManager, target, dirtyRecords, options)
+
+        // Syncing up records
+        for (i in 0 until totalSize) {
+            checkIfStopRequested()
+            val record = dirtyRecords[i]
+            val recordId = record.getString(SmartStore.SOUP_ENTRY_ID)
+            if (java.lang.Boolean.TRUE == recordIdToShouldSyncUp[recordId]) {
+                batch.add(record)
+            }
+
+            // Process batch if max batch size reached or at the end of dirtyRecordIds
+            if (batch.size == maxBatchSize || i == totalSize - 1) {
+                (target as AdvancedSyncUpTarget).syncUpRecords(
+                    syncManager,
+                    batch,
+                    options.fieldlist,
+                    options.mergeMode!!,
```

Comment:
I wasn't sure if this force-unwrap could be avoided.  It looks like this is a guard when deserializing `SyncOptions` from JSON in the event merge mode isn't specified.  Is there a reasonable default value? If it must be specified, could this operation or the creation of the sync skip this operation and log a warning?

Reply:
I've been getting rid of force unwrap as much as possible. I'll look at this one.

Reply:
Thanks.  If there's a logic alternative, great.  Perhaps if not, we move this to a guard with a clearer exception detail than the default provided by the force unwrap.  That could look something like `val mergeMode = options.mergeMode ?: throw IllegalArgumentException("Insert developer friendly message here.")`